### PR TITLE
feat: add effect size direction to coloc output

### PR DESCRIPTION
**Pull request by Tobi1kenobi**
## ✨ Context

Extracts the direction of effect from the study locus overlaps and appends it onto the colocalisation object.

## 🛠 What does this PR implement

Firstly adds a new method to StudyLocusOverlap to first calculate the ratio between each pair of left and right betas in each locus overlap. The sign of these betas is retained and then the averafe of these signs is returned as the output. See below diagram detailing process:
image

In theory this can return a value that is neither 1 or -1, however in practice this does not occur. See examples below when calculating overlaps between very large subsets of eQTL catalogue and a) Finngen
image
or b) GWAS catalog
image

Values between -1 and 1 could be captured by the front-end team if necessary and an acceptable threshold could be used e.g. beta_average_ratio > 0.8 = positive effect.

These values calculated and joined onto the colocalisation object.

## 🙈 Missing

No extensive testing of edge cases, wasn't sure how best to implement and the testing for colocs isn't particularly extensive as it is so didn't want to add a lot of extra.

## 🚦 Before submitting

- [x] Do these changes cover one single feature (one change at a time)?
- [x] Did you read the [contributor guideline](https://opentargets.github.io/gentropy/development/contributing/#contributing-checklist)?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you make sure there is no commented out code in this PR?
- [x] Did you follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standards in PR title and commit messages?
- [x] Did you make sure the branch is up-to-date with the `dev` branch?
- [x] Did you write any new necessary tests?
- [x] Did you make sure the changes pass local tests (`make test`)?
- [x] Did you make sure the changes pass pre-commit rules (e.g `poetry run pre-commit run --all-files`)?

### CODE REVIEW

**Comment by d0choa** [comment]
I added @DSuveges for review as well as there have been clashing changes coming from him

**Comment by addramir** [comment]
Please add fillterning by 0, otherwise is good for me.

**Comment by addramir** [comment]
@DSuveges  Have you had a chance to take a look?

File: src/gentropy/dataset/study_locus_overlap.py [hunk 1]
```diff
@@ -48,6 +51,36 @@ def from_associations(
...
 
+
+    def calculate_beta_ratio(self: StudyLocusOverlap) -> DataFrame:
+        """Calculate the beta ratio for the overlapping signals.
+
+        Returns:
+            DataFrame: A dataframe containing left and right loci IDs, chromosome
+            and the average sign of the beta ratio
+        """
+        expanded_overlaps = self.df.select("*", "statistics.*").drop("statistics")
```
Comment (DSuveges) [inline src/gentropy/dataset/study_locus_overlap.py:62]: I don't have objection about the logic in this function, however it is written in a quite different way than the other bits of the code. I would suggest collapsing the expressions as follows:

```python
return (
    # Unpack statistics column:
    self.df.select("*", "statistics.*")
    .drop("statistics")
    # Drop any rows where the beta is null or zero
    .filter(
        f.col("left_beta").isNotNull() & 
        f.col("right_beta").isNotNull() &
        (f.col("left_beta") != 0) &
        (f.col("right_beta") != 0) 
    )
    # Calculate the beta ratio and get the sign, then calculate the average sign across all variants in the locus
    .withColumn(
        "betaRatioSign",
        f.signum(f.col("left_beta") / f.col("right_beta"))
    )
    # Aggregate beta signs:
    .groupBy("leftStudyLocusId","rightStudyLocusId","chromosome")
    .agg(
        f.avg("betaRatioSign").alias("betaRatioSignAverage")
    )
)
```

File: src/gentropy/method/colocalisation.py [hunk 2]
```diff
@@ -179,6 +179,11 @@ def colocalise(
                     f.sum(f.col("clpp")).alias("clpp"),
                 )
                 .withColumn("colocalisationMethod", f.lit(cls.METHOD_NAME))
+                .join(overlapping_signals.calculate_beta_ratio(),
+                 ["leftStudyLocusId",
+                  "rightStudyLocusId",
+                   "chromosome"],
+                "left")
```
Comment (DSuveges) [inline src/gentropy/method/colocalisation.py:186]: I know, these are positional arguments, but I like to be more explicit in joins by providing 

```python
.join(
    overlapping_signals.calculate_beta_ratio(),
    on=["leftStudyLocusId", "rightStudyLocusId","chromosome"],
    how="left"
)
```

File: src/gentropy/method/colocalisation.py [hunk 3]
```diff
@@ -379,6 +384,11 @@ def colocalise(
                     "lH4bf",
                 )
                 .withColumn("colocalisationMethod", f.lit(cls.METHOD_NAME))
+                .join(overlapping_signals.calculate_beta_ratio(),
```
Comment (DSuveges) [inline src/gentropy/method/colocalisation.py:387]: Same here.

**Review by DSuveges — CHANGES_REQUESTED** [review]
I don't really have objection with the logic. It looks good and makes sense. However I would recommend to apply small stylistic updates just to make the code style more aligned with the rest of the codebase.

**Review by DSuveges — APPROVED** [review]
Thanks. Looks good now.

**Pull request merged**